src/Microsoft.ApiDesignGuidelines.Analyzers/Core/DoNotDecreaseInheritedMemberVisibility.cs

```diff
@@ -24,16 +27,81 @@ public abstract class DoNotDecreaseInheritedMemberVisibilityAnalyzer : Diagnosti
+                                                                             isEnabledByDefault: true,
                                                                              description: s_localizableDescription,
-                                                                             helpLinkUri: null,     // TODO: add MSDN url
+                                                                             helpLinkUri: "https://msdn.microsoft.com/en-us/library/ms182332.aspx",
                                                                              customTags: WellKnownDiagnosticTags.Telemetry);
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
 
         public override void Initialize(AnalysisContext analysisContext)
         { 
-            
+            analysisContext.RegisterSymbolAction(CheckForDecreasedVisibility, SymbolKind.Method, SymbolKind.Field, SymbolKind.Property, SymbolKind.Event);
+        }
+
+        private void CheckForDecreasedVisibility(SymbolAnalysisContext context)
+        {
+            // Bail out if the member is publicly accessible, or sealed, or on a sealed type
+            if (IsVisibleOutsideAssembly(context.Symbol) || context.Symbol.IsSealed || (context.Symbol.ContainingType?.IsSealed ?? true))
+            {
+                return;
+            }
+
+            // Find members on base types that share the member's name
+            var ancestorTypes = context.Symbol?.ContainingType?.GetBaseTypes() ?? Enumerable.Empty<INamedTypeSymbol>();
+            var hiddenOrOverriddenMembers = ancestorTypes.SelectMany(t => t.GetMembers(context.Symbol.Name));
+
+            if (hiddenOrOverriddenMembers.Any(IsVisibleOutsideAssembly))
+            {
+                context.ReportDiagnostic(context.Symbol.CreateDiagnostic(Rule));
+            }
+        }
+        
+        private bool IsVisibleOutsideAssembly(ISymbol symbol)
+        {
+            // If the containing type is not visible, then neither is the contained symbol
+            if (symbol.ContainingType != null && !IsVisibleOutsideAssembly(symbol.ContainingType))
+            {
+                return false;
+            }
+
+            // Public symbols are visible outside the assembly
+            if (symbol.DeclaredAccessibility == Accessibility.Public)
+            {
+                return true;
+            }
+
+            // Protected members are visible if on unsealed types
+            if ((symbol.DeclaredAccessibility == Accessibility.Protected || symbol.DeclaredAccessibility == Accessibility.ProtectedOrInternal)
+                && (symbol.ContainingType?.IsSealed == false))
+            {
+                return true;
+            }
+
+            // Check for explicit interface implementations if the symbol is a method, property, or event
+            if (symbol is IMethodSymbol)
```

Comment:
Instead of casting twice, consider:
```
if ((symbol as IMethodSymbol)?.ExplicitInterfaceImplementations.Any(IsVisibleOutsideAssembly) ?? false)
{
    return true;
}
```